File: Programming_2/C#/lab_7/LinkedList.cs
namespace lab_7
{
    public class LinkedList
    {
        internal Node head;
        
        internal class Node {  
            internal int data;  
            internal Node next; 
            
            public Node(int d) {  
                data = d;  
                next = null;  
            }  
        }

        public void CreateNode(int value)
        {
            head = new Node(value);
        }

        public void AddLast(int value)
        {
            if (head == null)
            {
                CreateNode(value);
                return;
            }

            var current = head.next;
            
            while (current != null)
            {
                current = current.next;
            }
            
            current = new Node(value);
        }
        
        public void AddAfterFirst(int value)
        {
            if (head == null)
            {
                CreateNode(value);
                return;
            }

            if (head.next == null)
            {
                head.next = new Node(value);
                return;
            }
            
            var temp = head.next;

            head.next = new Node(value) {next = temp};
        }
        
        public int Count
        {
            get
            {
                if (head == null) return -1;

                var current = head.next;
                var counter = 0;
                
                while (current != null)
                {
                    if (current.data % 2 == 0) ++counter;
                    if (current.next == null) break;
                    current = current.next.next;
                }

                return counter;
            }
        }
        
        public int IndexMinValue
        {
            get
            {
                if (head == null) return -1;

                var current = head.next;
                var minIndex = 0;
                var minValue = head.data;
                var loopCounter = 1;

                while (current != null)
                {
                    if (minValue > current.data)
                    {
                        minIndex = loopCounter;
                        minValue = current.data;
                    }

                    ++loopCounter;
                    current = current.next;
                }

                return minIndex;
            }
        }

        public void RemoveAfter(int index)
        {
            if (head == null) return;
            
            var current = head.next;
            
            var loopCounter = 0;
            while (current != null)
            {
                loopCounter++;
                if (loopCounter == index)
                {
                    current.next = null;
                    break;
                }
                current = current.next;
            }
        }
    }
}

File: Programming_2/C++/lab_7/LinkedList.cpp
//
// Created by Danil on 16.05.2020.
//

#include "LinkedList.h"

void LinkedList::CreateNode(int value) {
    node = new Node;
    node->value = value;
    node->next = nullptr;
}

void LinkedList::AddLast(int value) {
    Node *current = node;
    if (node == nullptr)
    {
        node = new Node{
                .value = value,
                .next = nullptr
        };
        return;
    }
    while (current->next) current = current->next;
    current->next = new Node{
            .value = value,
            .next = nullptr,
    };

}

void LinkedList::AddAfterFirst(int value) {
    if (node == nullptr)
    {
        node = new Node{
            .value = value,
            .next = nullptr
        };
        return;
    }
    else if (node->next == nullptr)
    {
        node->next = new Node{
                .value = value,
                .next = nullptr,
        };

        return;
    }
    Node *temp = node->next;
    node->next = new Node{
            .value = value,
            .next = temp,
    };
}

int LinkedList::Count() {
    Node *current = node->next;
    if (node == nullptr)
    {
        return -1;
    }
    int counter = 0;
    while (current != nullptr)
    {
        if (current->value % 2 == 0) ++counter;
        if (current->next == nullptr) break;
        current = current->next->next;

    }
    return counter;
}

int LinkedList::GetIndexMin() {
    Node *current = node->next;
    if (node == nullptr)
    {
        return -1;
    }
    int indexMin = 0;
    int min = node->value;
    int loopCounter = 1;
    while (current != nullptr)
    {
        if (current->value < min)
        {
            min = current->value;
            indexMin = loopCounter;
        }
        loopCounter++;
        current = current->next;

    }
    return indexMin;
}

void LinkedList::RemoveAfter(int index) {
    int loopCounter = -1;
    Node *current = node;
    while (current != nullptr)
    {
        loopCounter++;
        if (loopCounter == index) {
            current->next = nullptr;
            break;
        }
        current = current->next;
    }
}

LinkedList::~LinkedList() {
    RemoveAfter(0);
    delete node;
}

